src/Skia/Avalonia.Skia/DrawingContextImpl.cs

```diff
@@ -79,363 +138,554 @@ public void DrawGeometry(IBrush brush, Pen pen, IGeometryImpl geometry)
+            _currentOpacity *= opacity;
+        }
 
-            public void AddDisposable(IDisposable disposable)
-            {
-                if (_disposable1 == null)
-                    _disposable1 = disposable;
-                else if (_disposable2 == null)
-                    _disposable2 = disposable;
-                else
-                    throw new InvalidOperationException();
-            }
+        /// <inheritdoc />
+        public void PopOpacity()
+        {
+            _currentOpacity = _opacityStack.Pop();
+        }
 
-            public PaintWrapper(SKPaint paint)
+        /// <inheritdoc />
+        public virtual void Dispose()
+        {
+            if (_disposables == null)
             {
-                Paint = paint;
-                _disposable1 = null;
-                _disposable2 = null;
+                return;
             }
 
-            public void Dispose()
+            foreach (var disposable in _disposables)
             {
-                Paint?.Dispose();
-                _disposable1?.Dispose();
-                _disposable2?.Dispose();
+                disposable?.Dispose();
             }
         }
 
-        internal PaintWrapper CreatePaint(IBrush brush, Size targetSize)
+        /// <inheritdoc />
+        public void PushGeometryClip(IGeometryImpl clip)
         {
-            SKPaint paint = new SKPaint();
-            var rv = new PaintWrapper(paint);
-            paint.IsStroke = false;
+            Canvas.Save();
+            Canvas.ClipPath(((StreamGeometryImpl) clip).EffectivePath);
+        }
 
-            
-            double opacity = brush.Opacity * _currentOpacity;
-            paint.IsAntialias = true;
+        /// <inheritdoc />
+        public void PopGeometryClip()
+        {
+            Canvas.Restore();
+        }
 
-            var solid = brush as ISolidColorBrush;
-            if (solid != null)
+        /// <inheritdoc />
+        public void PushOpacityMask(IBrush mask, Rect bounds)
+        {
+            // TODO: This should be disposed
+            var paint = new SKPaint();
+
+            Canvas.SaveLayer(paint);
+            _maskStack.Push(CreatePaint(mask, bounds.Size));
+        }
+
+        /// <inheritdoc />
+        public void PopOpacityMask()
+        {
+            // TODO: This should be disposed
+            var paint = new SKPaint
             {
-                paint.Color = new SKColor(solid.Color.R, solid.Color.G, solid.Color.B, (byte) (solid.Color.A * opacity));
-                return rv;
+                BlendMode = SKBlendMode.DstIn
+            };
+
+            Canvas.SaveLayer(paint);
+            using (var paintWrapper = _maskStack.Pop())
+            {
+                Canvas.DrawPaint(paintWrapper.Paint);
             }
-            paint.Color = (new SKColor(255, 255, 255, (byte)(255 * opacity)));
 
-            var gradient = brush as IGradientBrush;
-            if (gradient != null)
+            Canvas.Restore();
```

Comment:
Double call Canvas.Restore is misprint?

Reply:
I don't think so. There's a Canvas.SaveLayer() call in PushOpacityMask that I think the second Canvas.Restore() corresponds to.

Reply:
@nc4rillo can you place a comment  for doc purposes? :)

Reply:
Sorry, I don't really understand why method need be called twice in succession! :C If it's alright, it will be great, if you'll left comment for people like me :)

Reply:
You’re right. EffectivePath is on the base class. No need to cast.

Reply:
I will add a comment to make things clearer.

Reply:
I wrapped the SKPaint in a using statement which had the effect of making it clear that Restore corresponds to the inner SaveLayer call.